File: neetcode150/18_bit_manipulation/0191_number_of_1_bits.cpp
/*
 * =====================================================================
 * LeetCode 191 : Number of 1 Bits                                   (Easy)
 * https://leetcode.com/problems/number-of-1-bits/
 * Category   : Bit Manipulation
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Given an unsigned 32-bit integer, return the number of '1' bits
 *   in its binary representation (the Hamming weight / popcount).
 *   Example: 11 = 1011₂ → 3.
 * ---------------------------------------------------------------------
 * INTUITION
 *   Counting needs a way to "remove one set bit per step". The trick
 *   x & (x - 1) clears exactly the LOWEST set bit of x — repeatedly
 *   applying it makes x zero in exactly `popcount(x)` steps, so the
 *   number of rounds IS the answer. No bit scanning, no wasted
 *   iterations over zero bits.
 * ---------------------------------------------------------------------
 * APPROACH 1 — Kernighan's algorithm (BEST)
 *   count = 0; while (n) { n &= n - 1; ++count; }. One round per
 *   SET bit — zeros never cost an iteration.
 *   Time  Complexity : O(popcount) ≤ O(32) = O(1).
 *   Space Complexity : O(1).
 * ---------------------------------------------------------------------
 * APPROACH 2 — Scan all 32 bits
 *   count = 0; for i in 0..31: count += n & 1; n >>= 1. Checks every
 *   position, including zeros — simple, predictable 32 rounds.
 *   Time  Complexity : O(32) = O(1).   Space Complexity : O(1).
 * ---------------------------------------------------------------------
 * DEEP DIVE — deriving x & (x - 1): why it clears the LOWEST 1-bit
 *   Write x in binary: ...1 0 0 0, where the lowest 1-bit sits at
 *   position k with only 0-bits below it. Subtracting 1 borrows
 *   through those zeros, flipping the bit at k to 0 and every bit
 *   below it to 1:
 *       x     = ...1 0 0 0  (bits above k unchanged)
 *       x - 1 = ...0 1 1 1  (bit k cleared, below becomes all-ones)
 *   Now AND them: positions above k match (both have the original
 *   bits), position k is 1 & 0 = 0, positions below are 0 & 1 = 0.
 *   Result: x with its lowest 1-bit removed — exactly one bit
 *   cleared, everything else untouched.
 * =====================================================================
 */

#include <iostream>      // std::cout — printing benchmark results
#include <vector>        // std::vector — the test-case table
#include <cstdint>       // std::uint32_t — the 32-bit operand type
#include <climits>       // UINT32_MAX — edge case for the test table
#include <chrono>        // std::chrono::steady_clock — timing for benchmark
#include <sys/resource.h>// getrusage — peak RAM measurement for benchmark

/* =====================================================================
 * APPROACH 1 : Kernighan — one iteration per SET bit
 * ===================================================================== */
class Solution_1 {
public:
    /*
     * Purpose : Count the set bits (Hamming weight) of a uint32_t.
     * Inputs  : n — the value to count bits of.
     * Output  : int — the number of 1-bits in n.
     */
    int hammingWeight(std::uint32_t n) {
        int count = 0;

        // Each round wipes exactly ONE set bit (the lowest one), so
        // the loop runs popcount(n) times — zero bits cost nothing.
        while (n != 0) {
            n &= n - 1;   // clears the lowest set bit (see deep dive)
            ++count;
        }
        return count;
    }
};

/* =====================================================================
 * APPROACH 2 : Scan all 32 bit positions
 * ===================================================================== */
class Solution_2 {
public:
    /*
     * Purpose : Count the set bits by examining every one of the 32
     *           positions, including the zero bits.
     * Inputs  : n — the value to count bits of.
     * Output  : int — the number of 1-bits in n.
     */
    int hammingWeight(std::uint32_t n) {
        int count = 0;

        // Fixed 32 rounds: pull the lowest bit off, add it, shift.
        // n >>= 1 on an UNSIGNED int is a logical shift — zero fills
        // the top, so the loop runs exactly 32 times and stops.
        for (int i = 0; i < 32; ++i) {
            count += n & 1;
            n >>= 1;
        }
        return count;
    }
};

/* =====================================================================
 * BENCHMARK — run both approaches and compare time + peak RAM
 * ===================================================================== */
int main() {
    // Test table: { input, expected popcount } — official example,
    // a sparse value, and the all-ones boundary.
    std::vector<std::pair<std::uint32_t, int>> tests = {
        {11u, 3},        // 1011₂ → three 1-bits
        {128u, 1},       // 10000000₂ → one 1-bit
        {UINT32_MAX, 32} // all ones → thirty-two 1-bits
    };

    // --- Approach 1: run every test, measure time + peak RAM ---
    struct rusage before_1, after_1;
    getrusage(RUSAGE_SELF, &before_1);
    auto start_1 = std::chrono::steady_clock::now();
    bool ok_1 = true;
    for (const auto& t : tests) {
        int got = Solution_1().hammingWeight(t.first);
        bool pass = got == t.second;
        ok_1 = ok_1 && pass;
        std::cout << "  A1 hammingWeight(" << t.first << ") = " << got
                  << "  expected " << t.second
                  << (pass ? "  PASS" : "  FAIL") << "\n";
    }
    auto end_1 = std::chrono::steady_clock::now();
    getrusage(RUSAGE_SELF, &after_1);

    // nanoseconds → microseconds; ru_maxrss is ALREADY in KB on Linux.
    double us_1 = std::chrono::duration_cast<std::chrono::nanoseconds>(end_1 - start_1).count() / 1000.0;
    long mem_1 = after_1.ru_maxrss - before_1.ru_maxrss;  // KB

    // --- Approach 2: run every test, measure time + peak RAM ---
    struct rusage before_2, after_2;
    getrusage(RUSAGE_SELF, &before_2);
    auto start_2 = std::chrono::steady_clock::now();
    bool ok_2 = true;
    for (const auto& t : tests) {
        int got = Solution_2().hammingWeight(t.first);
        bool pass = got == t.second;
        ok_2 = ok_2 && pass;
        std::cout << "  A2 hammingWeight(" << t.first << ") = " << got
                  << "  expected " << t.second
                  << (pass ? "  PASS" : "  FAIL") << "\n";
    }
    auto end_2 = std::chrono::steady_clock::now();
    getrusage(RUSAGE_SELF, &after_2);

    double us_2 = std::chrono::duration_cast<std::chrono::nanoseconds>(end_2 - start_2).count() / 1000.0;
    long mem_2 = after_2.ru_maxrss - before_2.ru_maxrss;  // KB

    // Summary lines: time in µs, memory delta in KB.
    std::cout << "Approach 1 (Kernighan)      : time = " << us_1
              << " µs  peak-ram-delta = " << mem_1 << " KB\n";
    std::cout << "Approach 2 (32-bit scan)    : time = " << us_2
              << " µs  peak-ram-delta = " << mem_2 << " KB\n";

    std::cout << (ok_1 && ok_2
                  ? "PASS : both approaches count every test value correctly."
                  : "FAIL : at least one test mismatched.")
              << "\n";

    return 0;
}

File: neetcode150/18_bit_manipulation/0191_number_of_1_bits.py
"""
=====================================================================
LeetCode 191 : Number of 1 Bits                                   (Easy)
https://leetcode.com/problems/number-of-1-bits/
Category   : Bit Manipulation
---------------------------------------------------------------------
PROBLEM
    Given an unsigned 32-bit integer, return the number of '1' bits
    in its binary representation (the Hamming weight / popcount).
    Example: 11 = 1011₂ → 3.
---------------------------------------------------------------------
INTUITION
    Counting needs a way to "remove one set bit per step". The trick
    x & (x - 1) clears exactly the LOWEST set bit of x — repeated
    application makes x zero in exactly popcount(x) steps, so the
    round count IS the answer. No iteration is wasted on zeros.
---------------------------------------------------------------------
APPROACH 1 — Kernighan's algorithm (BEST)
    count = 0; while n: n &= n - 1; count += 1. One round per SET
    bit. Derivation of x & (x - 1):
        x     = ...1 0 0 0   (lowest 1-bit at position k)
        x - 1 = ...0 1 1 1   (borrow: bit k → 0, below → all 1s)
        x & (x-1) = ...1·1 ··· 0 ··· 0·1 = x minus its lowest 1-bit
    Time  Complexity : O(popcount) ≤ O(32) = O(1)
    Space Complexity : O(1)
    PYTHON NOTE: n is unbounded — without a 32-bit mask, Kernighan
    also strips set bits above bit 31 (fine: they aren't in the
    input's 32-bit lane) but the count never exceeds 32 either way.
---------------------------------------------------------------------
APPROACH 2 — Scan all 32 bit positions
    count = 0; for i in range(32): count += n & 1; n >>= 1.
    Checks every position including zeros — simple, 32 fixed rounds.
    Time  Complexity : O(32) = O(1)   Space Complexity : O(1)
---------------------------------------------------------------------
REAL LIFE
    • Parity bits & error detection: popcount parity of a byte decides
      the stored parity bit in ECC memory and serial links.
    • Population statistics: IPv4 subnet masks, Hamming distance in
      similarity engines (Bitcoin mining chips count ones too).
=====================================================================
"""

from __future__ import annotations
import time
import tracemalloc
from typing import List, Tuple

MASK32 = 0xFFFFFFFF  # emulates the 32-bit hardware lane in Python


# =====================================================================
# APPROACH 1 : Kernighan — one iteration per SET bit (best)
# =====================================================================
class Solution_1:
    """
    Purpose : Count the set bits (Hamming weight) of a 32-bit value.
    Inputs  : n — the value to count bits of.
    Output  : int — the number of 1-bits in n (0..32).
    """

    def hammingWeight(self, n: int) -> int:
        count = 0

        # Each round wipes exactly ONE set bit (the lowest one), so
        # the loop runs popcount(n) times — zero bits cost nothing.
        # The mask keeps Python's unbounded int inside the 32-bit lane
        # so the loop provably terminates for any input.
        n &= MASK32
        while n != 0:
            n &= n - 1   # clears the lowest set bit (see deep dive)
            count += 1
        return count


# =====================================================================
# APPROACH 2 : Scan all 32 bit positions
# =====================================================================
class Solution_2:
    """
    Purpose : Count the set bits by examining every one of the 32
              positions, including the zero bits.
    Inputs  : n — the value to count bits of.
    Output  : int — the number of 1-bits in n (0..32).
    """

    def hammingWeight(self, n: int) -> int:
        count = 0

        # Fixed 32 rounds: pull the lowest bit off, add it, shift.
        # Masking after the shift keeps the emulated lane 32 bits wide.
        for _ in range(32):
            count += n & 1
            n = (n >> 1) & MASK32
        return count


# =====================================================================
# BENCHMARK — run both approaches and compare time + peak memory
# =====================================================================
if __name__ == "__main__":
    # Test table: {input, expected popcount} — official example,
    # a sparse value, and the all-ones boundary.
    tests: List[Tuple[int, int]] = [
        (11, 3),          # 1011₂ → three 1-bits
        (128, 1),         # 10000000₂ → one 1-bit
        (0xFFFFFFFF, 32), # all ones → thirty-two 1-bits
    ]

    # --- Approach 1: time + peak memory (tracemalloc snapshot) ---
    tracemalloc.start()
    t0 = time.perf_counter()
    ok_1 = True
    for inp, exp in tests:
        got = Solution_1().hammingWeight(inp)
        pass_ = got == exp
        ok_1 = ok_1 and pass_
        print(f"  A1 hammingWeight({inp}) = {got}  expected {exp}  {'PASS' if pass_ else 'FAIL'}")
    t1 = time.perf_counter()
    _, peak_1 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # --- Approach 2: time + peak memory ---
    tracemalloc.start()
    t0b = time.perf_counter()
    ok_2 = True
    for inp, exp in tests:
        got = Solution_2().hammingWeight(inp)
        pass_ = got == exp
        ok_2 = ok_2 and pass_
        print(f"  A2 hammingWeight({inp}) = {got}  expected {exp}  {'PASS' if pass_ else 'FAIL'}")
    t1b = time.perf_counter()
    _, peak_2 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # Summary lines: time in µs, peak memory in KB.
    print(f"Approach 1 (Kernighan)      : time = {(t1 - t0) * 1e6:.2f} µs  peak-mem = {peak_1 / 1024:.2f} KB")
    print(f"Approach 2 (32-bit scan)    : time = {(t1b - t0b) * 1e6:.2f} µs  peak-mem = {peak_2 / 1024:.2f} KB")

    print("PASS : both approaches count every test value correctly."
          if ok_1 and ok_2
          else "FAIL : at least one test mismatched.")
